server/aesdsocket.c

```diff
@@ -10,97 +10,400 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/stat.h>
+#include <sys/signal.h>
+#include <sys/time.h>
+#include "queue.h"
 #include <unistd.h>
 #include <netdb.h>
 #include <string.h>
 #include <stdlib.h>
+#include <time.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <syslog.h>
 #include <linux/fs.h>
 #include <arpa/inet.h>
 #include <netinet/in.h>
 #include <signal.h>
+#include <stdint.h>
+#include <pthread.h>
 
-#define BACKLOG 10
+//-----------------------Global--Defines-----------------------------
+#define BACKLOG (10)
+#define RECEIVE_BUFFER (1024)
+#define SEND_BUFFER (1024)
+#define TEMP_BUFFER (1024)
 
-const char *file = "/var/tmp/aesdsocketdata";
-int sockfd, new_conn_fd;
-char *buf;
-struct addrinfo *servinfo;
-struct sockaddr_storage new_conn;
-int fd;
+//--------------------Globals-Struct---------------------------------------
 
-void sig_handler(int signum)
+typedef struct thread_data
 {
-    freeaddrinfo(servinfo);
-    close(sockfd);
-    free(buf);
-    unlink(file);
-    exit(EXIT_SUCCESS);
-}
+    pthread_t threadID;
+    int conn_fd;
+    int thread_complete_flag;
+    SLIST_ENTRY(thread_data)
+    entries;
+} thread_data;
+
+struct globals_aesdsocket
+{
+    int sockfd;
+    int filefd;
+    int file_writehead;
+    int exit_flag;
+    int write_flag;
+    thread_data head;
+    pthread_mutex_t file_mutex;
+} aesdsocket;
+
+// Defining a struct with name slisthead
+SLIST_HEAD(slisthead, thread_data);
+
+//--------------------Function Declarations--------------------------------
+static void register_signal_handlers();
+static void bind_to_port(char *port);
+static void open_temp_file(char *file);
+static void daemonify(int argc, char *argv[]);
+static void accept_connections_loop(struct slisthead *head);
+static void write_to_file();
+static void send_file();
+static void write_timestamp();
+static void start_timer();
+void *thread_function(void *threadparams);
+
+// ---------------------------------main--------------------------------------------
 
 int main(int argc, char *argv[])
 {
+    // Init Logging
+    openlog(NULL, 0, LOG_USER);
+
+    register_signal_handlers();
+
+    bind_to_port("9000");
 
-    // Register Signal Handlers
-    if (signal(SIGINT, sig_handler) == SIG_ERR)
+    pthread_mutex_init(&aesdsocket.file_mutex, NULL);
+
+    daemonify(argc, argv);
+
+    open_temp_file("/var/tmp/aesdsocketdata");
+
+    start_timer();
+
+    struct slisthead head;
+
+    SLIST_INIT(&head);
+
+    while (1)
     {
+        if (aesdsocket.write_flag)
+        {
+            write_timestamp();
```

Comment:
is there locking before writing timestamp to file?

Reply:
Yes, if you follow the function it leads to a write_to_file function which has the locking